Lib/typing.py

```diff
@@ -1950,6 +1976,29 @@ class Film(TypedDict):
     return isinstance(tp, _TypedDictMeta)
 
 
+def assert_never(arg: Never, /) -> Never:
+    """Statically assert that a line of code is unreachable.
+
+    Example::
+
+        def int_or_str(arg: int | str) -> None:
+            match arg:
+                case int():
+                    print("It's an int")
+                case str():
+                    print("It's a str")
+                case _:
+                    assert_never(arg)
+
+    If a type checker finds that a call to assert_never() is
+    reachable, it will emit an error.
+
+    At runtime, this throws an exception when called.
+
+    """
+    raise RuntimeError("Expected code to be unreachable")
```

Comment:
Has it already been litigated whether `AssertionError` is a better exception to raise here?

Reply:
Hasn't been discussed yet. Personally I don't like raising AssertionError manually; I feel it should only be used from the actual `assert` statement.

Reply:
I quite like the idea of using `AssertionError` here, actually. `assert`s and `AssertionError`s are for places in code that you hope will never be reached, if everything's working the way you expected it to. If you hit an `AssertionError`, you know there's been a logical error in the code that was written. And that feels very fitting for this situation, whereas a `RuntimeError` could be caused by all sorts of things.

You could always use `assert False` if you don't like the way `raise AssertionError` _looks_ ;)

Reply:
The `assert` statement becomes a no-op if `__debug__` is False. Would it be better if `assert_never` does the same?

Reply:
I switched it to AssertionError.

I don't think making the behavior rely on `__debug__` makes sense here: that's a performance optimization to strip out assertions so you don't pay the cost of checking them, but if you hit an assert_never(), there's nothing else to do.

Reply:
Correct, when raising `AssertionError` manually you generally shouldn't check `__debug__`. (The latter exists so you can write code more elaborate than a bunch of `assert` statements without having to turn it into a function.)